### PR TITLE
Adds custom expiration time based on grant type

### PR DESCRIPTION
**Pull request by rishabhsairawat**
### Summary
This PR aims to solve #996 **Expiration Time Base On Grant Type**

### CODE REVIEW

**Comment by nbulaj** [comment]
Hi @rishabhsairawat . Thank you for th PR!

I think that we can extract grant type names (like `'client_credentials'` and other) in some constant like:

```ruby
module Doorkeeper 
  module OAuth
    FLOWS = [
      CLIENT_CREDENTIALS = 'client_credentials'.freeze,
      REFRESH_TOKEN = 'refresh_token'.freeze,
      # ...
    ].freeze
  
    # ...
  end
end
```

File: lib/doorkeeper/oauth/base_request.rb [hunk 1]
```diff
@@ -2,6 +2,8 @@ module Doorkeeper
   module OAuth
     class BaseRequest
       include Validations
+      
+      attr_accessor :grant_type
```
Comment (nbulaj) [inline lib/doorkeeper/oauth/base_request.rb]: Do we really need accessor ? I think `attr_reader` would be enough

File: lib/doorkeeper/oauth/authorization/token.rb [hunk 2]
```diff
@@ -36,7 +36,7 @@ def issue_token
             pre_auth.client,
             resource_owner.id,
             pre_auth.scopes,
-            self.class.access_token_expires_in(configuration, pre_auth),
+            self.class.access_token_expires_in(configuration, pre_auth, 'implicit'),
```
Comment (nbulaj) [inline lib/doorkeeper/oauth/authorization/token.rb]: Use constants for such things as I described in the topic comments.

File: spec/controllers/authorizations_controller_spec.rb [hunk 3]
```diff
@@ -30,10 +30,13 @@ def translated_error_message(key)
   let(:client)        { FactoryBot.create :application }
   let(:user)          { User.create!(name: 'Joe', password: 'sekret') }
   let(:access_token)  { FactoryBot.build :access_token, resource_owner_id: user.id, application_id: client.id }
-
+  
   before do
     allow(Doorkeeper.configuration).to receive(:grant_flows).and_return(["implicit"])
     allow(controller).to receive(:current_resource_owner).and_return(user)
+    allow(Doorkeeper.configuration).to receive(:custom_access_token_expires_in).and_return(proc {|app, grant|
+        grant == 'implicit' ?  1234 : nil
```
Comment (nbulaj) [inline spec/controllers/authorizations_controller_spec.rb]: Use flows constants in specs too.

File: lib/doorkeeper/oauth/authorization/token.rb (hunk 2 above)
Comment (rishabhsairawat) [reply lib/doorkeeper/oauth/authorization/token.rb]: Hi @nbulaj Where to define these constants for grant flows as there is no class for grants in Doorkeeper::OAuth module. Should I define them  in BaseRequest class

Comment (nbulaj) [reply lib/doorkeeper/oauth/authorization/token.rb]: Let's place them into `Doorkeeper::Request` module as `GRANT_TYPES`

Comment (nbulaj) [reply lib/doorkeeper/oauth/authorization/token.rb]: No, let's introduce  `oauth.rb` with explicit OAuth module:

```ruby
module Doorkeeper
  module OAuth
    GRANT_TYPES = [
     # ...
    ]. freeze
  end
end
```

Comment (rishabhsairawat) [reply lib/doorkeeper/oauth/authorization/token.rb]: I was thinking the same as It would be better to place them in separate file.

**Comment by nbulaj** [comment]
Thank you @rishabhsairawat 👍

**Pull request merged**